File: week_1__C/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

long get_size(void);
int numbers(long n);
void card_company(long n);

int main(void) {
  long n = get_size();

  if (numbers(n) == 1) { // check if credit card number is valid
    card_company(n);     // if valid, identify issuing credit card company
  }

  return 0;
}

long get_size(void) {
  long n;

  printf("\n");

  n = get_long("Please enter a credit card number here: \n");

  return n;
}

int numbers(long n) {
  int digit;
  int position = 0;
  int sum_doubles = 0;
  int sum_rest = 0;
  int sum;

  while (n > 0) {
    digit = n % 10; // get the last digit
    n /= 10;        // remove the last digit

    // 2nd to last and so forth:
    if (position % 2 != 0) {
      digit = digit * 2;
      sum_doubles += digit % 10 + digit / 10;
    }
    // last and so forth:
    else if (position % 2 == 0) {
      sum_rest += digit;
    }
    position++;
  }

  sum = sum_doubles + sum_rest;

  if (sum % 10 == 0) {
    return 1;
  } else {
    printf("INVALID\n");
    return 0;
  }
}

void card_company(long n) {
  // American Express: 15 digits, starts with 34 or 37
  // MasterCard: 16 digits, starts with 51, 52, 53, 54, or 55
  // Visa: 13 or 16 digits, starts with 4

  if (n >= 340000000000000 && n <= 349999999999999 ||
      n >= 370000000000000 && n <= 379999999999999) {
    printf("AMEX\n");

  } else if (n >= 5100000000000000 && n <= 5599999999999999) {
    printf("MASTERCARD\n");

  } else if (n >= 4000000000000 && n <= 4999999999999 ||
             n >= 4000000000000000 && n <= 4999999999999999) {
    printf("VISA\n");
  } else {
    printf("INVALID\n");
  }
}

File: week_6__python/pset6/sentimental-credit/credit.py
def main():

    card_number = get_size()
    numbers(card_number)
    card_company(card_number)
    

def get_size():
    card_number = int(input("Number to check, please: "))
    return card_number


def numbers(card_number):
    # Hans Peter Luhn's algorithm:
    card_number = list(str(card_number))
    
    # turn them into a list of numbers:
    card_number = [int(i) for i in card_number]

    # reverse the list because we need to start from the second to last from the right:
    card_number.reverse()

    # make a new list and multiply every other number by two:
    temporary_list = [int(digit) * 2 for digit in card_number if digit % 2 != 0]

    # split every number that is > 9 into two digits:
    temporary_list = [int(digit) for digit in temporary_list for digit in str(digit)]

    # make a new list and add the digits together:
    sum_temporary_list = sum(temporary_list)

    # make a new list and add the digits that weren't multiplied by 2 together:
    rest_list = [int(digit) for digit in card_number if digit % 2 == 0]

    # sum the rest_list:
    sum_rest_list = sum(rest_list)
    
    # 2. Add the sum to the sum of the digits that weren’t multiplied by 2.
    total = sum_temporary_list + sum_rest_list
    
    # 3. If the total’s last digit is 0 (or, put more formally, if the total modulo 10 is congruent to 0), the number is valid!
    if total % 10 == 0:
        print("VALID\n")
        return True
    else:
        print("INVALID\n")
        return False
    
    
def card_company(card_number):
    if card_number >= 340000000000000 and card_number <= 349999999999999 or card_number >= 370000000000000 and card_number <= 379999999999999:
        print("AMEX\n")
    elif card_number >= 5100000000000000 and card_number <= 5599999999999999:
        print("MASTERCARD\n")
    elif card_number >= 4000000000000 and card_number <= 4999999999999 or card_number >= 4000000000000000 and card_number <= 4999999999999999:
        print("VISA\n")
    else:
        return False


main()